### PR TITLE
Add tests for Defaults

### LINKED ISSUE #375: Atomic defaults tests
**Summary**

The kuadrant operator now allows defaults to be set for AuthPolicy and RateLimitPolicy. Test that these defaults work properly and are mutually exclusive with setting policy in spec.

**Documentation**

Implementing PRs: https://github.com/Kuadrant/kuadrant-operator/pull/503, https://github.com/Kuadrant/kuadrant-operator/pull/456, https://github.com/Kuadrant/kuadrant-operator/pull/536
RFC: https://github.com/Kuadrant/architecture/blob/main/rfcs/0009-defaults-and-overrides.md

**Test cases**

Repeat each for AuthPolicy and RateLimitPolicy

* Tests functionality of rules set within the `defaults` block
    - Set HTTPRoute attached policy with an explicit defaults block
    - verify the policy inside the defaults block functions as a normal policy
 
* Test mutual exclusivity of `defaults` block with other rules
    - set HTTPRoute attached policy configured both with regular config and config inside a `defaults` block
    - Check that policy is rejected due to `Implicit and explicit defaults are mutually exclusive`

* Test policy inheritability and PolicyAffected condition
    -  Attach a policy with a configured `defaults` block to your gateway
    - Check the status of the gateway's HTTPRoute
    - status should specify `Object affected by {Auth|RateLimit}Policy x`
    - Curl the route to verify the policy attached to gateway got applied to your HTTPRoute

* Test policy discoverability
    -  Attach a policy with a configured `defaults` block to your gateway
    - Create a new HTTPRoute attached to your gateway
    - Check HTTPRoute status, it should specify `Object affected by {Auth|RateLimit}Policy x`
    - Curl the route to verify the policy attached to gateway got applied to your HTTPRoute

### PR DESCRIPTION
**Pull request by averevki**
### Changes

- Add the Defaults&Overrides interface to the testsuite
- Add tests for the Defaults

Closes #375

### CODE REVIEW

**Comment by martinhesko** [comment]
Already reviewed in the previous PR, LGTM

**Review by martinhesko — APPROVED** [review]

**Comment by averevki** [comment]
signoff commits^

**Review by martinhesko — DISMISSED** [review]

File: testsuite/tests/singlecluster/authorino/authorization/opa/external_registry/conftest.py [hunk 1]
```diff
@@ -24,5 +24,5 @@ def authorization(authorization, opa_policy_expectation):
     Adds OPA policy. Rego query is located on external registry (Mockserver).
     Policy accepts requests that contain `header`.
     """
-    authorization.authorization.add_external_opa_policy("opa", opa_policy_expectation, 1)
+    authorization.rules.authorization.add_external_opa_policy("opa", opa_policy_expectation, 1)
```
Comment (pehala) [inline testsuite/tests/singlecluster/authorino/authorization/opa/external_registry/conftest.py:27]: I would make it backwards compatible  and make `rules` default if nothing is specified. Will result in much smaller PR without any significant sideeffect

File: testsuite/kuadrant/policy/authorization/auth_policy.py [hunk 2]
```diff
@@ -46,3 +45,24 @@ def add_rule(self, when: list["Rule"]):
...
+    def defaults(self):
+        """Add new rule into the `defaults` AuthPolicy section"""
+        self.spec_section = self.model.spec.setdefault("defaults", {})
+        return self
+
+    @property
+    def overrides(self):
+        """Add new rule into the `overrides` AuthPolicy section"""
+        self.spec_section = self.model.spec.setdefault("overrides", {})
+        return self
```
Comment (pehala) [inline testsuite/kuadrant/policy/authorization/auth_policy.py:68]: This is weird, while it indeed works, wont this cause weird artefacts like this?

```
authorization.identity.oidc() ->rules
authorization.overrides.identity.oidc() -> overrides
authorization.identity.oidc() -> overrides
```

Comment (averevki) [reply testsuite/kuadrant/policy/authorization/auth_policy.py:68]: spec_section resets after new item is added. I have similar setup below 
```py
authorization.defaults.identity.add_oidc("inside-defaults", oidc_provider.well_known["issuer"])
authorization.identity.add_oidc("outside-defaults", oidc_provider.well_known["issuer"])
```

Comment (pehala) [reply testsuite/kuadrant/policy/authorization/auth_policy.py:68]: That is awesome, thanks!

File: testsuite/tests/singlecluster/conftest.py [hunk 3]
```diff
@@ -34,10 +34,12 @@ def authorization_name(blame):
 
 
 @pytest.fixture(scope="module")
-def authorization(kuadrant, route, authorization_name, cluster, label):
+def authorization(request, kuadrant, route, gateway, blame, cluster, label):  # pylint: disable=unused-argument
     """Authorization object (In case of Kuadrant AuthPolicy)"""
+    target_ref = request.getfixturevalue(getattr(request, "param", "route"))
```
Comment (pehala) [inline testsuite/tests/singlecluster/conftest.py:39]: I do not like that indirect parametrization is so high in the hierarchy. It probably wont affect anything, but I still do not like it. @jsmolar thoughts?

File: testsuite/tests/singlecluster/defaults/test_rules_exclusivity.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+
+@pytest.mark.parametrize(
+    "component_fixture",
+    [
+        pytest.param("authorization", id="AuthPolicy"),
+        pytest.param("rate_limit", id="RateLimitPolicy"),
+    ],
+)
+def test_rules_exclusivity(request, component_fixture):
```
Comment (pehala) [inline testsuite/tests/singlecluster/defaults/test_rules_exclusivity.py:40]: This is too confusing. I would rather haev two test cases with creation of the policies inside it as the act of creation is part of the testcase and as such should be done in test function and not in fixtures

Comment (pehala) [reply testsuite/tests/singlecluster/defaults/test_rules_exclusivity.py:40]: Difference between this test and most other tests is that most other tests expect policy to exist and only test behaviour afterwards, hence creation is not important and is not part of the testfunction. This test tests the creation itself and as such is much more similar to `test_gw_doesnt_exists` and similar reconciliation tests. In these test the creation of policy itself is the tested behaviour and should be present in the test function and not in the fixture, making it hard to understand what is going on for someone who is not familiar with pytest internals.

TLDR: For this test I think it would be better to 1) simplify it 2) not use fixtures for policies and instead create them directly in the test function

Comment (averevki) [reply testsuite/tests/singlecluster/defaults/test_rules_exclusivity.py:40]: Makes sense. Thanks for the explanation

File: testsuite/tests/singlecluster/conftest.py (hunk 3 above)
Comment (jsmolar) [reply testsuite/tests/singlecluster/conftest.py:39]: I added exactly the same parametrization to rate limit tests and it is opaque shit. I don't know how to solve parametrization for routes and gateways, so if you have better idea I'm all ears, but if not lets go with this solution.

**Review by pehala — APPROVED** [review]

**Review by martinhesko — APPROVED** [review]
